internal/filesystem/nfs/client.go

```diff
@@ -0,0 +1,611 @@
+package nfs
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/djdv/go-filesystem-utils/internal/filesystem"
+	fserrors "github.com/djdv/go-filesystem-utils/internal/filesystem/errors"
+	"github.com/djdv/go-filesystem-utils/internal/generic"
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/willscott/go-nfs-client/nfs"
+	"github.com/willscott/go-nfs-client/nfs/rpc"
+)
+
+type (
+	goFS struct {
+		target        *nfs.Target
+		linkSeparator string
+		linkLimit     uint
+		// NOTE [2024.01.08]: The NFS server library is able to handle multiple requests concurrently
+		// but the client library is not intended to handle multiple outstanding operations.
+		// As a result, we lock on each operation.
+		// If this changes upstream, we can drop the operation mutex.
+		opMu sync.Mutex
+	}
+	goShared struct {
+		opMu    *sync.Mutex
+		netName string
+	}
+	goFile struct {
+		file *nfs.File
+		goShared
+	}
+	goDirectory struct {
+		goShared
+		target  *nfs.Target
+		entries []*nfs.EntryPlus
+	}
+	goEnt struct {
+		*nfs.EntryPlus
+	}
+	// goStatWrapper wraps the [fs.FileInfo] returned from
+	// [nfs.target.Getattr] (which returns an empty name
+	// in its `.Name` method).
+	goStatWrapper struct {
+		fs.FileInfo
+		name string
+	}
+	clientSettings struct {
+		*goFS
+		hostname, dirpath      string
+		uid, gid               uint32
+		defaultUID, defaultGID bool
+	}
+	ClientOption func(*clientSettings) error
+)
+
+const errStale = generic.ConstError("handle became stale")
+
+func NewNFSGuest(maddr multiaddr.Multiaddr, options ...ClientOption) (*goFS, error) {
+	var (
+		fsys = goFS{
+			linkLimit: 40, // Arbitrary; Linux's default.
+		}
+		settings = makeClientSettings(&fsys)
+	)
+	if err := generic.ApplyOptions(&settings, options...); err != nil {
+		return nil, err
+	}
+	if err := settings.fillDefaults(); err != nil {
+		return nil, err
+	}
+	naddr, err := manet.ToNetAddr(maddr)
+	if err != nil {
+		return nil, err
+	}
+	const (
+		network    = "tcp"
+		privledged = false
```

Comment:
```suggestion
		privileged = false
```

Reply:
(I'll probably end up squashing this before merge and just attribute you as Co-Author on the whole client commit.)

Reply:
lol, I didn't update any references to this. I'll fix it.